src/loop/parser.ml

```diff
@@ -190,70 +192,131 @@ module Make(State : State.S) = struct
-                ~language:lang (`Raw (filename, lang, contents)) in
-            let file = { file with loc = file_loc; lang = Some lang; } in
-            st, file, gen_finally gen cl
-          | Some `Full ->
-            let lang, file_loc, l = Logic.parse_raw_lazy ~language:lang
-              ~filename contents
-            in
-            let file = { file with loc = file_loc; lang = Some lang; } in
-            st, file, gen_of_llist l
-          end
-        | `File f ->
-          let s = Dolmen.Std.Statement.include_ f [] in
-          (* Auto-detect input format *)
-          let lang =
-            match file.lang with
-            | Some l -> l
-            | None ->
-              let res, _, _ = Logic.of_filename f in
-              res
-          in
-          (* Formats Dimacs and Tptp are descriptive and lack the emission
-              of formal solve/prove instructions, so we need to add them. *)
-          let s' =
-            match lang with
-            | Logic.Zf
-            | Logic.ICNF
-            | Logic.Smtlib2 _
-            | Logic.Alt_ergo -> s
-            | Logic.Dimacs
-            | Logic.Tptp _ ->
-              Dolmen.Std.Statement.pack [s; Dolmen.Std.Statement.prove ()]
-          in
-          let file = { file with lang = Some lang; } in
-          st, file, (Gen.singleton s')
-      with
-      | Logic.Extension_not_found ext ->
-        State.error st extension_not_found ext, file, Gen.empty
+    match source with
+    | `Raw (filename, contents) ->
+      begin match file.mode with
+      | None
+      | Some `Incremental ->
+        let lang, file_loc, gen, cl = Logic.parse_input
+            ~language:lang (`Raw (filename, lang, contents)) in
+        let file = { file with loc = file_loc; lang = Some lang; } in
+        st, file, gen_finally gen cl
+      | Some `Full ->
+        let lang, file_loc, l = Logic.parse_raw_lazy ~language:lang
+          ~filename contents
+        in
+        let file = { file with loc = file_loc; lang = Some lang; } in
+        (* We are called from inside a [gen'] so the returned generator is
+        always evaluated immediately, there is no need to add laziness *)
+        st, file, Gen.of_list (Lazy.force l)
+      end
+    | `File f ->
+      let s = Dolmen.Std.Statement.include_ f [] in
+      (* Formats Dimacs and Tptp are descriptive and lack the emission
+          of formal solve/prove instructions, so we need to add them. *)
+      let s' =
+        match lang with
+        | Logic.Zf
+        | Logic.ICNF
+        | Logic.Smtlib2 _
+        | Logic.Alt_ergo -> s
+        | Logic.Dimacs
+        | Logic.Tptp _ ->
+          Dolmen.Std.Statement.pack [s; Dolmen.Std.Statement.prove ()]
+      in
+      let file = { file with lang = Some lang; } in
+      st, file, (Gen.singleton s')
+
+  (* This is a ['a Gen.t] with a threaded state. *)
+  type 'a gen' = {g: State.t -> State.t * 'a option} [@@unboxed]
+
+  let make g = {g}
+
+  let run {g} = g
```

Comment:
are the `make` and `run` functions really necessary here ?

Reply:
Not really, I just don't like "exposing" `{g}`. I can remove them.